artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java

```diff
@@ -0,0 +1,166 @@
+
+public class RingQueue extends QueueImpl {
+
+   private static final Logger logger = Logger.getLogger(RingQueue.class);
+   private volatile long ringSize;
+
+   public RingQueue(final long persistenceID,
+                    final SimpleString address,
+                    final SimpleString name,
+                    final Filter filter,
+                    final PageSubscription pageSubscription,
+                    final SimpleString user,
+                    final boolean durable,
+                    final boolean temporary,
+                    final boolean autoCreated,
+                    final RoutingType routingType,
+                    final Integer maxConsumers,
+                    final Boolean exclusive,
+                    final Boolean groupRebalance,
+                    final Integer groupBuckets,
+                    final SimpleString groupFirstKey,
+                    final Integer consumersBeforeDispatch,
+                    final Long delayBeforeDispatch,
+                    final Boolean purgeOnNoConsumers,
+                    final Long ringSize,
+                    final Boolean nonDestructive,
+                    final Boolean autoDelete,
+                    final Long autoDeleteDelay,
+                    final Long autoDeleteMessageCount,
+                    final boolean configurationManaged,
+                    final ScheduledExecutorService scheduledExecutor,
+                    final PostOffice postOffice,
+                    final StorageManager storageManager,
+                    final HierarchicalRepository<AddressSettings> addressSettingsRepository,
+                    final ArtemisExecutor executor,
+                    final ActiveMQServer server,
+                    final QueueFactory factory) {
+      super(persistenceID, address, name, filter, pageSubscription, user, durable, temporary, autoCreated, routingType, maxConsumers, exclusive, groupRebalance, groupBuckets, groupFirstKey, nonDestructive, consumersBeforeDispatch, delayBeforeDispatch, purgeOnNoConsumers, autoDelete, autoDeleteDelay, autoDeleteMessageCount, configurationManaged, scheduledExecutor, postOffice, storageManager, addressSettingsRepository, executor, server, factory);
+      this.ringSize = ringSize;
+   }
+
+   @Override
+   public synchronized void addTail(final MessageReference ref, final boolean direct) {
+      enforceRing();
+
+      super.addTail(ref, direct);
+   }
+
+   @Override
+   public synchronized void addHead(final MessageReference ref, boolean scheduling) {
+      enforceRing(ref, scheduling);
+
+      if (!ref.isAlreadyAcked()) {
+         super.addHead(ref, scheduling);
+      }
+   }
+
+   @Override
+   public boolean allowsReferenceCallback() {
+      return false;
+   }
+
+   private void enforceRing() {
+      enforceRing(null, false);
+   }
+
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
```

Comment:
I think we should judge whether ref is already acked first.

E.g. for non destructive queue, suppose ring size is 1, m1 is sent and delivered to consumer c1. Then m2 is sent and delivered to consumer c2. Thus m1 is replace acked and m2 resides in "messageReferences"(message count is now 1). Now consumer c1 is closed without acking m1, m1 is rollbacked and replace acked again that would throw IllegalStateException.